src/probing.rs

```diff
@@ -0,0 +1,154 @@
+	target_cache_reuse_limit: usize,
+	targets: Mutex<Vec<PublicKey>>,
+	target_index: Mutex<usize>,
+	uses_since_load: Mutex<usize>,
+}
+
+impl HighCapacityStrategy {
+	pub fn new(network_graph: Arc<Graph>, max_targets: usize, max_reloads: usize) -> Self {
+		Self {
+			network_graph,
+			max_targets_per_cycle: max_targets,
+			target_cache_reuse_limit: max_reloads,
+			targets: Mutex::new(Vec::new()),
+			target_index: Mutex::new(0),
+			uses_since_load: Mutex::new(0),
+		}
+	}
+
+	fn network_graph(&self) -> NetworkGraph {
+		NetworkGraph::new(Arc::clone(&self.network_graph))
+	}
+}
+
+impl ProbingStrategy for HighCapacityStrategy {
+	fn load_targets(&self) {
+		let mut targets = self.targets.lock().unwrap();
+		let mut uses_since_load = self.uses_since_load.lock().unwrap();
+
+		if !targets.is_empty() && *uses_since_load < self.target_cache_reuse_limit {
+			*uses_since_load += 1;
+			return;
+		}
+
+		let network = self.network_graph();
+		let highest_capacity_nodes =
+			network.select_highest_capacity_nodes(self.max_targets_per_cycle);
+		*targets =
+			highest_capacity_nodes.iter().filter_map(|node_id| node_id.as_pubkey().ok()).collect();
+
+		let mut target_index = self.target_index.lock().unwrap();
+		*target_index = 0;
+		*uses_since_load = 0;
+	}
+
+	fn next_target(&self) -> Option<PublicKey> {
+		let mut target_index = self.target_index.lock().unwrap();
+		let targets = self.targets.lock().unwrap();
+
+		if *target_index < targets.len() {
+			let pk = targets[*target_index];
+			*target_index += 1;
+			Some(pk)
+		} else {
+			// reset index for next cycle
+			*target_index = 0;
+			None
+		}
+	}
+}
+
+/// Configuration for the probing service used to evaluate channel liquidity by sending pre-flight
+/// probes to peers and routes.
+pub struct ProbingService {
+	pub probing_interval_secs: u64,
+	probing_amount_msat: u64,
+	strategy: Arc<dyn ProbingStrategy + Send + Sync>,
+	config: Arc<Config>,
+	logger: Arc<Logger>,
+	channel_manager: Arc<ChannelManager>,
+	keys_manager: Arc<KeysManager>,
+	is_running: Arc<RwLock<bool>>,
+	payment_store: Arc<PaymentStore>,
+}
+
+impl ProbingService {
+	/// Creates a new probing service with the given configuration and dependencies.
+	pub fn new(
+		probing_interval_secs: u64, probing_amount_msat: u64,
+		strategy: Arc<dyn ProbingStrategy + Send + Sync>, config: Arc<Config>, logger: Arc<Logger>,
+		channel_manager: Arc<ChannelManager>, keys_manager: Arc<KeysManager>,
+		is_running: Arc<RwLock<bool>>, payment_store: Arc<PaymentStore>,
+	) -> Self {
+		Self {
+			probing_interval_secs,
+			probing_amount_msat,
+			strategy,
+			config,
+			logger,
+			channel_manager,
+			keys_manager,
+			is_running,
+			payment_store,
+		}
+	}
+
+	pub fn handle_probing(&self) {
+		self.strategy.load_targets();
+		loop {
+			if let Some(target) = self.strategy.next_target() {
+				let spontaneous_payment = self.spontaneous_payment();
+				match spontaneous_payment.send_probes(self.probing_amount_msat, target) {
```

Comment:
We probably want to be setting https://docs.rs/lightning/0.2.0-rc1/lightning/routing/scoring/struct.ProbabilisticScoringFeeParameters.html#structfield.probing_diversity_penalty_msat to something non-0 when doing background probing. Sadly the LDK `ChannelManager::send_spontaneous_preflight_probes` API used under the hood here doesn't allow overriding the scoring parameters, but calling `ChannelManager:send_probe` directly should be pretty easy, I think.

Reply:
Yes, I agree that this code should probably use the 'raw' LDK API here rather than reusing the LDK Node APIs. We might want to expose the diversity penalty, as well as the liquidity multiplier from `send_preflight_probes` (https://github.com/lightningdevkit/rust-lightning/blob/7177cfc958bd54e12fff730d337630131679b066/lightning/src/ln/channelmanager.rs#L5971-L6049) to be used.